File: src/deuce/EmacsLispError.java
package deuce;

import clojure.lang.Symbol;

public class EmacsLispError extends RuntimeException {
    public final Symbol symbol;
    public final Object data;

    public EmacsLispError(Object data) {
        this(null, data);
    }

    public EmacsLispError(Symbol symbol, Object data) {
        this.data = data;
        this.symbol = symbol;
    }

    public String getMessage() {
        return String.format("Lisp error: (no-catch %s %s)", symbol, data);
    }

    public String toString() {
        return getMessage();
    }
}

File: src/deuce/EmacsLispError.clj
(ns deuce.EmacsLispError
  (:gen-class :extends RuntimeException
              :init init
              :constructors {[Object] []
                             [Object clojure.lang.Symbol] []}
              :state state))

(defn -init [data & [symbol]]
  [[] {:data data :symbol symbol}])

(defn -getMessage [this]
  (if-let [data (:data (.state this))]
    (format "Lisp error: (%s)" data)
    (format "Lisp error: [no-catch %s nil]" symbol)))

(defn -toString [this]
  (.getMessage this))
